firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/Chunks.cpp

```diff
@@ -0,0 +1,427 @@
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus input_status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        input_status = chunks.input(val);
+        THEN("the input status is equal to ok") {
+          REQUIRE(input_status == PF::Protocols::ChunkInputStatus::ok);
+        }
+      }
+
+      PF::Util::Vector<char, buffer_size> buffer;
+      PF::Protocols::ChunkOutputStatus output_status = chunks.output(buffer);
+      THEN("the output status is equal to waiting") {
+        REQUIRE(output_status == PF::Protocols::ChunkOutputStatus::waiting);
+      }
+      THEN("the output buffer is empty") {
+        REQUIRE(buffer.empty() == true);
+      }
+
+      auto final_status = chunks.input(val);
+      THEN("the final status is equal to invalid_length") {
+        REQUIRE(final_status == PF::Protocols::ChunkInputStatus::invalid_length);
+      }
+    }
+  }
+}
+
+SCENARIO(
+    "Protocols::ChunkSplitter behaves correctly across chunk boundary(256 bytes)", "[chunks]") {
+  GIVEN("A uint8_t chunk splitter of internal buffer equal to 256 bytes") {
+    constexpr size_t buffer_size = 256;
+    PF::Protocols::ChunkSplitter<buffer_size, uint8_t> chunks;
+
+    WHEN("enough data is input to exactly fill the buffer of the splitter") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+        THEN("the final status is equal to ok") {
+          REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        }
+      }
+    }
+
+    WHEN("the input data exceeds bounds without output being called") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+        THEN("the initial status is equal to ok") {
+          REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        }
+      }
+
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+        THEN("the final status is equal to invalid_length") {
+          REQUIRE(status == PF::Protocols::ChunkInputStatus::invalid_length);
+        }
+      }
+
+    }
+
+    WHEN("the input data exceeds bounds after output is called, without delimiter") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+        THEN("the initial status is equal to ok") {
+          REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        }
+      }
+
+      PF::Util::Vector<uint8_t, buffer_size> buffer;
+      PF::Protocols::ChunkOutputStatus output_status = chunks.output(buffer);
+      THEN("the output status is equal to waiting") {
+        REQUIRE(output_status == PF::Protocols::ChunkOutputStatus::waiting);
+      }
+
+      auto final_status = chunks.input(val);
```

Comment:
I previously suggested trying out putting these code lines between the `THEN`s, and now I see that it can make things more confusing. I think in this case we should move line 106 to between lines 101 and 102, because it almost looks as if we need an "AND WHEN" when we have this line between the THENs, since we're calling the `input` method which can change the state of `chunks`.

Reply:
Fixed.